core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java

```diff
@@ -517,18 +517,12 @@ protected QueryEvaluationStep prepare(Service service, QueryEvaluationContext co
 	}
 
 	protected QueryEvaluationStep prepare(Filter node, QueryEvaluationContext context) throws QueryEvaluationException {
-
-		if (FilterIterator.isPartOfSubQuery(node)) {
-			context = new FilterIterator.RetainedVariableFilteredQueryEvaluationContext(node, context);
-		}
```

Comment:
This is the change that fixes the issue. It looks like the original reason for this code was to handle some legacy stuff, I presume something related to `CompareAll` and `CompareAny` since these classes are not in use anymore.

Reply:
Here is the code for `FilterIterator.isPartOfSubQuery(...)`:
```
	public static boolean isPartOfSubQuery(QueryModelNode node) {
		if (node instanceof SubQueryValueOperator) {
			return true;
		}

		QueryModelNode parent = node.getParentNode();
		if (parent == null) {
			return false;
		} else {
			return isPartOfSubQuery(parent);
		}
	}
```

Reply:
`SubQueryValueOperator` is subclasses by `In`, `Exists`, `CompareAll` and `CompareAny`. It is also subclassed by a the abstract class `CompareSubQueryValueOperator`, which seems insignificant.